embassy-mspm0/src/uart.rs

```diff
@@ -0,0 +1,1085 @@
+            let Some((ibrd, fbrd)) = calculate_brd(clock, div, baudrate, oversampling) else {
+                continue;
+            };
+
+            if ibrd < MIN_IBRD || fbrd > MAX_FBRD {
+                continue;
+            }
+
+            found = Some((hse_value, div_value, ibrd, fbrd));
+        }
+    }
+
+    let Some((hse, div, ibrd, fbrd)) = found else {
+        return Err(ConfigError::InvalidBaudRate);
+    };
+
+    regs.clkdiv().write(|w| {
+        w.set_ratio(div);
+    });
+
+    regs.ibrd().write(|w| {
+        w.set_divint(ibrd);
+    });
+
+    regs.fbrd().write(|w| {
+        w.set_divfrac(fbrd);
+    });
+
+    regs.ctl0().modify(|w| {
+        w.set_hse(hse);
+    });
+
+    Ok(())
+}
+
+/// Calculate the integer and fractional parts of the `BRD` value.
+///
+/// Returns [`None`] if calculating this results in overflows.
+///
+/// Values returned are `(ibrd, fbrd)`
+fn calculate_brd(clock: u32, div: u8, baud: u32, oversampling: u8) -> Option<(u16, u8)> {
+    use fixed::types::U26F6;
+
+    // Calculate BRD according to SLAU 846 section 18.2.3.4.
+    //
+    // BRD is a 22-bit value with 16 integer bits and 6 fractional bits.
+    //
+    // uart_clock = clock / div
+    // brd = ibrd.fbrd = uart_clock / (oversampling * baud)"
+    //
+    // It is tempting to rearrange the equation such that there is only a single division in
+    // order to reduce error. However this is wrong since the denominator ends up being too
+    // small to represent in 6 fraction bits. This means that FBRD would always be 0.
+    //
+    // Calculations are done in a U16F6 format. However the fixed crate has no such representation.
+    // U26F6 is used since it has the same number of fractional bits and we verify at the end that
+    // the integer part did not overflow.
+    let clock = U26F6::from_num(clock);
+    let div = U26F6::from_num(div);
+    let oversampling = U26F6::from_num(oversampling);
+    let baud = U26F6::from_num(baud);
+
+    let uart_clock = clock.checked_div(div)?;
+
+    // oversampling * baud
+    let denom = oversampling.checked_mul(baud)?;
+    // uart_clock / (oversampling * baud)
+    let brd = uart_clock.checked_div(denom)?;
+
+    // Checked is used to determine overflow in the 10 most singificant bits since the
+    // actual representation of BRD is U16F6.
+    let ibrd = brd.checked_to_num::<u16>()?;
+
+    // We need to scale FBRD's representation to an integer.
+    let fbrd_scale = U26F6::from_num(2_u32.checked_pow(U26F6::FRAC_NBITS)?);
+
+    // It is suggested that 0.5 is added to ensure that any fractional parts round up to the next
+    // integer. If it doesn't round up then it'll get discarded which is okay.
+    let half = U26F6::from_num(1) / U26F6::from_num(2);
+    // fbrd = INT(((FRAC(BRD) * 64) + 0.5))
+    let fbrd = brd
+        .frac()
+        .checked_mul(fbrd_scale)?
+        .checked_add(half)?
+        .checked_to_num::<u8>()?;
+
+    Some((ibrd, fbrd))
+}
```

Comment:
The last thing before I merge. I would like someone to double check the calculations here.

[18.2.3.4 Baud Rate Generation (page 1351)](https://www.ti.com/lit/ug/slau846/slau846.pdf) describes the calculations for `IBRD` and `FBRD`. There is an example at the bottom of this file which does verify the example in the datasheet is produced. This also does work since the examples at 115200 baud for the virtual com port do work.

Reply:
@korbin I gave this an attempt to make it const and it would involve manual `match`/`if let`/`let` guard statements in a lot of places. I feel that makes it far harder to read and understand. Until `impl const` is stable I doubt this function will be easier as const.